akka-actor-typed/src/main/scala/akka/actor/typed/eventstream/EventStream.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.eventstream
+
+import akka.actor.typed._
+import akka.actor.typed.internal.adapter.EventStreamAdapter
+import akka.actor.typed.scaladsl.adapter._
+import akka.annotation.{ DoNotInherit, InternalApi }
+
+import scala.reflect.ClassTag
+
+/**
+ * INTERNAL API
+ *
+ * Exposes a typed actor that interacts with the [[akka.actor.ActorSystem.eventStream]].
+ *
+ * It is used as an extension to ensure a single instance per actor system.
+ */
+@InternalApi private[akka] final class EventStream(actorSystem: ActorSystem[_]) extends Extension {
+  val ref: ActorRef[EventStream.Command] =
+    actorSystem.internalSystemActorOf(EventStreamAdapter.behavior, "eventstream", Props.empty)
+}
+
+object EventStream extends ExtensionId[EventStream] {
+
+  /**
+   * Not for user Extension
+   */
+  @DoNotInherit sealed trait Command
+
+  /**
+   * Publish an event of type E
+   * @param event
+   * @tparam E
+   */
+  final case class Publish[E](event: E) extends Command
+
+  /**
+   * Subscribe a typed actor to listen for types or subtypes of E.
+   * ==Simple example==
+   * {{{
+   *   sealed trait A
+   *   case object A1 extends A
+   *   //listen for all As
+   *   def subscribe(actorSystem: ActorSystem[_], actorRef: ActorRef[A]) =
+   *     actorSystem.eventStream ! Subscribe(actorRef)
+   *   //listen for A1s only
+   *   def subscribe(actorSystem: ActorSystem[_], actorRef: ActorRef[A]) =
+   *     actorSystem.eventStream ! Subscribe[A1](actorRef)
+   * }}}
+   *
+   * @param subscriber
+   * @param classTag
+   * @tparam E
+   */
+  final case class Subscribe[E](subscriber: ActorRef[E])(implicit classTag: ClassTag[E]) extends Command {
+    private[akka] def topic: Class[_] = classTag.runtimeClass
+  }
+
+  /**
+   * Unsubscribe an actor ref from the event stream
+   * @param subscriber
+   * @tparam E
+   */
+  final case class Unsubscribe[E](subscriber: ActorRef[E]) extends Command
+
+  override def createExtension(system: ActorSystem[_]): EventStream = new EventStream(system)
```

Comment:
it's somewhat strange to have `object EventStream` as public but the returned `class EventStream` is `private[akka]`. The extension can be completely internal and placed in `akka.actor.typed.internal.EventStreamExt`. Keeping the `object EventStream` here for defining the messages.

Reply:
Hm, can I do the inverse? In my opinion is strange for a user to import `akka.actor.typed.internal.EventStreamExt`
So I suggest instead I do the EventStream class EventStreamInternal . What do you think?

Reply:
Ah sorry, totally misread the ext as external, working on it now

Reply:
I named `EventStreamExtension` instead of Ext, let me know if that's ok